pgml-extension/pgml_rust/src/bindings/linfa.rs

```diff
@@ -69,49 +69,103 @@ impl Bindings for LinearRegression {
     }
 }
 
-// #[derive(Debug)]
-// pub struct LogisticRegression {
-//     estimator: linfa_logistic::FittedLogisticRegression<f32, f32>,
-//     num_features: usize,
-// }
-
-// impl Bindings for LogisticRegression {
-//     fn fit(dataset: &Dataset) -> Box<dyn Bindings> where Self: Sized {
-//         let estimator = linfa_logistic::LogisticRegression::default();
-//         let estimator = estimator.fit(&dataset.into()).unwrap();
-
-//         Box::new(LogisticRegression { estimator, num_features: dataset.num_features })
-//     }
-
-//     /// Predict a novel datapoint.
-//     fn predict(&self, features: &[f32]) -> f32 {
-//         self.predict_batch(features)[0]
-//     }
-
-//     /// Predict a novel datapoint.
-//     fn predict_batch(&self, features: &[f32]) -> Vec<f32> {
-//         let records = ArrayView2::from_shape(
-//             (features.len() / self.num_features, self.num_features),
-//             features,
-//         )
-//         .unwrap();
-//         self.estimator.predict(records).targets.into_raw_vec()
-//     }
-
-//     /// Deserialize self from bytes, with additional context
-//     fn from_bytes(bytes: &[u8]) -> Box<dyn Bindings> where Self: Sized {
-//         let estimator: LogisticRegression = rmp_serde::from_read(bytes).unwrap();
-//         Box::new(estimator)
-//     }
-
-//     /// Serialize self to bytes
-//     fn to_bytes(&self) -> Vec<u8> {
-//         // Logistic Regression doesn't support Serde (or any kind of ) serialization
-//         todo!()
-//     }
-
-//     /// The hyperparams used during the fit call
-//     fn hyperparams(&self) -> Hyperparams {
-//         todo!()
-//     }
-// }
+#[derive(Debug, Serialize, Deserialize)]
+pub struct LogisticRegression {
+    estimator_binary: Option<linfa_logistic::FittedLogisticRegression<f32, i32>>,
+    estimator_multi: Option<linfa_logistic::MultiFittedLogisticRegression<f32, i32>>,
+    num_features: usize,
+    num_distinct_labels: usize,
+}
+
+impl LogisticRegression {
+    pub fn fit(dataset: &Dataset, _hyperparams: &Hyperparams) -> Box<dyn Bindings>
+    where
+        Self: Sized,
+    {
+        let records = ArrayView2::from_shape(
+            (dataset.num_train_rows, dataset.num_features),
+            &dataset.x_train,
+        )
+        .unwrap();
+
+        // Copy to convert to i32 because LogisticRegression doesn't contineous targets.
+        let y_train: Vec<i32> = dataset.y_train.iter().map(|x| *x as i32).collect();
```

Comment:
If you use |&x| you wont have to dereference... tomaeto/tomahto.

Reply:
Clever.